### ISSUE TITLE
write_log_message api should unescape message

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#12393]
See https://github.com/scylladb/scylladb/blob/8797958dfce783d6fd02a41bf1b32d3171eb5573/api/system.cc#L67

The message is passed as a parameter in the http request, hence is expected to be "escaped" so it won't contain white spaces, or special characters like `[&<>]`.

for example: `scylla-api-client system log POST --message "hello world" --level warn` printed the following to the log:
```
WARN  2022-12-23 13:00:24,285 [shard 0] api - /system/log: hello+world
```

**Comment by VOID001** [comment]
Then if the parameter is a url-encoded text such as `%2B`, should the log printout literally `%2B` or `+`?

**Comment by bhalevy** [comment]
The latter. Scylla should perform url decoding and print `+` rather than `%2B`

**Comment by VOID001** [comment]
Thanks for the reply.
After I tried this on my side, I found that the parameter seems already get decoded by seastar `parse
_query_param`. On line 54 in seastar/src/http/request.cc

Is this really a bug or it already works as
 expected?

**Comment by bhalevy** [comment]
you're right, we are unquoting `%` escaped codes.
Apparently, on the api client side, `requests.post` uses something like https://docs.python.org/3/library/urllib.parse.html#urllib.parse.quote_plus
that quotes spaces as `+` characters so we need to decode them in addition to the `%` escape codes.

**Comment by VOID001** [comment]
May I ask where I can get the client you mentioned in the issue? I didn't find it in this repo

**Comment by nyh** [comment]
I don't think there's a bug in Scylla - it just uses the correct Seastar API `req.get_query_param("message")` which is supposed to do the necessary decoding. As @VOID001 noted above it already does URL decoding but... there is indeed a mess with the "+": The URI RFC specifies that a space should be encoded as `%20`. **not** a `+`.  The `+` is an HTML (!) invention, that has no place in REST API that has nothing to do with HTML.

Still, if you're wondering what we can do, we can't do it in Scylla, we need to do this in Seastar - decoding the "+" to a space needs to happen **before** the percent decoding, otherwise, when the client encodes a "+" to `%2B` if we first decoded it to + and then to space, it will be wrong. So the "+" decoding needs to happen inside `get_query_param()`, it can't happen after it returns.

I suggest you should close this issue and then either fix the client, or, if you think we may face similar broken clients in the future - create a Seastar issue to change the decoding.

**Comment by bhalevy** [comment]
@nyh agreed.
I think the simplest solution would be to decode the plus character to space in seastar. Fixing the scylla-api-client around https://github.com/scylladb/scylla-api-client/blob/485b25d95df0e8c56c1639a8a8be99b6c676e799/scylla_api_client/rest/__init__.py#L96 may be possible too but it seems it will require using a different library than `requests` to prevent this encoding of space to a plus char.

**Comment by nyh** [comment]
Just one correction: If we open a Seastar issue we shouldn't close this one yet - it should be closed only by the patch updating the Seastar module with the fix. Ideally we should also have a test for this in Scylla.

**Comment by VOID001** [comment]
Sounds good! I can work on this unit test and fix-up (if the issue is going to be fixed in Scylla). The `requests` library is widely used in Python, it seems that this library is not posting the data via URI, but via

**Comment by VOID001** [comment]
@bhalevy I saw the MR is already submitted to seastar. Do we need to add an unit test to test this case in scylla as well? I want know if I can take this as a `good-first-issue` to start developing on scylla.

**Comment by nyh** [comment]
Fixed in Seastar: https://github.com/scylladb/seastar/pull/1531
When we update the Seastar submodule in Scylla, we can close this issue.

@VOID001 yes, it would be nice to have a unit test for this issue. A good place to put this test is test/rest-api/, but on the other hand to check that the logging actually logged the space and not the "+", you'll need to read the log file. I have existing code for that in test/cql-pytest/test_logs.py - so maybe you should put the new tests in that file, not in rest-api/. To write to the log you can add in test/cql-pytest/rest_api.py a simple function like scylla_log() from test/alternator/util.py. On the other hand, I don't know how we can be sure that the sending function really sent a "+" for the space - if it doesn't we won't be testing the new fix :-(

**Comment by VOID001** [comment]
Can we add a unit test for `get_query_params` and test with '+' in the post data. Instead of do it end to end, we can provide exact the '+' as param to the http `parse_query` part. Therefore we don't need to worry about whether the python script actually send a '+' instead of a ' ' or '%20'

**Comment by bhalevy** [comment]
> Can we add a unit test for `get_query_params` and test with '+' in the post data. Instead of do it end to end, we can provide exact the '+' as param to the http `parse_query` part. Therefore we don't need to worry about whether the python script actually send a '+' instead of a ' ' or '%20'

You mean in seastar?
In any case, feel free to post a PR with your proposed test so the discussion would be about a concrete change.

**Comment by bhalevy** [comment]
Fixed in 4a18ee87eb8b7ca037ad148bd3c43b7a32cb0b74 that updated the seastar submodule

**Comment by bhalevy** [comment]
fix exists in 5.4, but not in 5.2
since it's cosmetic not backport is required
